File: app/src/main/java/com/ndmsystems/coala/layers/arq/ArqLayer.java
package com.ndmsystems.coala.layers.arq;

import com.ndmsystems.coala.BuildConfig;
import com.ndmsystems.coala.CoAPClient;
import com.ndmsystems.coala.CoAPHandler;
import com.ndmsystems.coala.CoAPMessagePool;
import com.ndmsystems.coala.helpers.Hex;
import com.ndmsystems.coala.helpers.MessageHelper;
import com.ndmsystems.coala.layers.ReceiveLayer;
import com.ndmsystems.coala.layers.SendLayer;
import com.ndmsystems.coala.layers.arq.data.DataFactory;
import com.ndmsystems.coala.layers.arq.states.LoggableState;
import com.ndmsystems.coala.layers.arq.states.ReceiveState;
import com.ndmsystems.coala.layers.arq.states.SendState;
import com.ndmsystems.coala.message.CoAPMessage;
import com.ndmsystems.coala.message.CoAPMessageCode;
import com.ndmsystems.coala.message.CoAPMessageOption;
import com.ndmsystems.coala.message.CoAPMessageOptionCode;
import com.ndmsystems.coala.message.CoAPMessagePayload;
import com.ndmsystems.coala.message.CoAPMessageType;
import com.ndmsystems.coala.utils.Reference;
import com.ndmsystems.infrastructure.logging.LogHelper;

import net.jodah.expiringmap.ExpirationPolicy;
import net.jodah.expiringmap.ExpiringMap;

import java.net.InetSocketAddress;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.TimeUnit;

/**
 * Created by bas on 02.08.17.
 */

public class ArqLayer implements ReceiveLayer, SendLayer {

    private static final int WINDOW_SIZE = 70;
    private static final int MAX_PAYLOAD_SIZE = 1024;

    private final CoAPClient client;
    private final CoAPMessagePool messagePool;

    private final Map<String, ReceiveState> receiveStates = ExpiringMap.builder()
            .expirationPolicy(ExpirationPolicy.ACCESSED)
            .expiration(10, TimeUnit.SECONDS)
        .build();
    private final Map<String, SendState> sendStates = new ConcurrentHashMap<>();

    public ArqLayer(CoAPClient client,
                    CoAPMessagePool messagePool) {
        this.client = client;
        this.messagePool = messagePool;
    }

    private boolean isBlockedMessage(CoAPMessage message) {
        return message.hasOption(CoAPMessageOptionCode.OptionBlock1) ||
                message.hasOption(CoAPMessageOptionCode.OptionBlock2);
    }

    private boolean isAboutArq(CoAPMessage message) {
        return message.hasOption(CoAPMessageOptionCode.OptionSelectiveRepeatWindowSize);
    }

    private Block getBlock1(CoAPMessage message) {
        CoAPMessagePayload payload = message.getPayload();
        byte[] data = payload == null ? null : message.getPayload().content;
        return new Block((int) message.getOption(CoAPMessageOptionCode.OptionBlock1).value,
                data);
    }

    private Block getBlock2(CoAPMessage message) {
        CoAPMessagePayload payload = message.getPayload();
        byte[] data = payload == null ? null : message.getPayload().content;
        return new Block((int) message.getOption(CoAPMessageOptionCode.OptionBlock2).value,
                data);
    }

    @Override
    public boolean onReceive(CoAPMessage message, Reference<InetSocketAddress> senderAddressReference) {

        if (!isAboutArq(message)) {
            return true;
        }

        if (message.getCode() == CoAPMessageCode.CoapCodeEmpty
                && message.getType() == CoAPMessageType.ACK && !message.hasOption(CoAPMessageOptionCode.OptionBlock1)) {//For block1 block2 mixed mode, message ending block1 start block2, like Received data from Peer, id 51486, payload:'', address: 138.197.191.160:5683 type: ACK code: CoapCodeEmpty path: null schema: coap token 951e67cdc72af162
            //Options: OptionBlock1 : 'com.ndmsystems.coala.layers.arq.Block@a046fab'(8174) OptionSelectiveRepeatWindowSize : '70', address: 138.197.191.160:5683 type: CON code: CoapCodeContent path: null schema: coap token 951e67cdc72af162
            //Options: OptionBlock2 : 'com.ndmsystems.coala.layers.arq.Block@72bd608'(30) OptionSelectiveRepeatWindowSize : '300'
            messagePool.setNoNeededSending(message);
            return false;
        }

        if (!isBlockedMessage(message)) {
            return true;
        }

        if (message.getToken() == null) {
            sendResetMessage(message, senderAddressReference.get());
        }

        int windowSize = (int) message.getOption(CoAPMessageOptionCode.OptionSelectiveRepeatWindowSize).value;

        CoAPMessage ackMessage = CoAPMessage.ackTo(message, senderAddressReference.get(), CoAPMessageCode.CoapCodeEmpty);

        if (message.hasOption(CoAPMessageOptionCode.OptionBlock1)) {
            Block block = getBlock1(message);
            return process(message, CoAPMessageOptionCode.OptionBlock1, block, windowSize, ackMessage);
        }

        if (message.hasOption(CoAPMessageOptionCode.OptionBlock2)) {
            Block block = getBlock2(message);
            return process(message, CoAPMessageOptionCode.OptionBlock2, block, windowSize, ackMessage);
        }
        return true;
    }

    private void sendResetMessage(CoAPMessage incomingMessage, InetSocketAddress fromAddress) {
        CoAPMessage resetMessage = CoAPMessage.resetTo(incomingMessage, fromAddress);
        client.send(resetMessage, null);
    }

    private boolean process(CoAPMessage incomingMessage, CoAPMessageOptionCode blockOptionCode, Block block, int windowSize, CoAPMessage ackMessage) {
        String token = Hex.encodeHexString(incomingMessage.getToken());

        switch (incomingMessage.getType()) {
            case ACK:
            case RST:
                // Transmit ACK
                didTransmit(block.getNumber(), token);
                messagePool.remove(incomingMessage);
                SendState sendState = sendStates.get(token);
                if (sendState != null && sendState.isCompleted()) {
                    LogHelper.v("ARQ: Sending completed, pushing to message pool original message" + sendState.getOriginalMessage().getId());
                    CoAPMessage originalMessage = sendState.getOriginalMessage();
                    if (incomingMessage.getCode() == CoAPMessageCode.CoapCodeEmpty
                            && incomingMessage.getType() == CoAPMessageType.ACK) {
                        messagePool.add(originalMessage);
                        messagePool.setNoNeededSending(originalMessage);
                        sendStates.remove(token);
                        return false;

                    } else {
                        incomingMessage = originalMessage;
                        sendStates.remove(token);
                        return true;
                    }
                }
                break;
            case CON:
                // Receive CON
                CoAPMessagePayload payload = incomingMessage.getPayload();
                if (payload == null) {
                    LogHelper.e("ARQ: payload expected for token = " + token);
                    throw new RuntimeException("ARQ: payload expected for token = " + token);
                }

                ReceiveState receiveState = receiveStates.get(token);

                if (receiveState == null) {
                    LogHelper.v("ARQ: creating ReceiveState for token = " + token);
                    receiveState = new ReceiveState(incomingMessage);
                    receiveStates.put(token, receiveState);
                }

                receiveState.didReceiveBlock(block, incomingMessage.getCode());

                ackMessage.addOption(new CoAPMessageOption(blockOptionCode, block.toInt()));
                ackMessage.addOption(new CoAPMessageOption(CoAPMessageOptionCode.OptionSelectiveRepeatWindowSize, windowSize));

                if (receiveState.isTransferCompleted()) {
                    //receiveStates.remove(token);

                    CoAPMessage originalMessage = messagePool.getSourceMessageByToken(incomingMessage.getHexToken());
                    if (originalMessage != null) {
                        LogHelper.v("ARQ: Receive " + incomingMessage.getHexToken() + " completed, size " + receiveState.getDataSize()
                                + ", passing message " + originalMessage.getId() + ", with token: " + Hex.encodeHexString(originalMessage.getToken())
                                + " along");
                        if (originalMessage.hasOption(CoAPMessageOptionCode.OptionProxyURI)) {
                            ackMessage.addOption(originalMessage.getOption(CoAPMessageOptionCode.OptionProxyURI));
                        }
                        if (originalMessage.getProxy() != null) {
                            ackMessage.setProxy(originalMessage.getProxy());
                            ackMessage.setProxySecurityId(originalMessage.getProxySecurityId());
                        }
                        incomingMessage.setId(originalMessage.getId());
                    } else {
                        LogHelper.v("ARQ: Receive with originalMessage = null, token " + incomingMessage.getHexToken() + " completed, size "
                                + receiveState.getDataSize());
                    }
                    ackMessage.setCode(CoAPMessageCode.CoapCodeEmpty);
                    client.send(ackMessage, null);

                    //incomingMessage = messagePool.getSourceMessageByToken(incomingMessage.getHexToken());
                    incomingMessage.setPayload(new CoAPMessagePayload(receiveState.getData()));
                    incomingMessage.setCode(CoAPMessageCode.CoapCodeContent);
                    incomingMessage.setType(CoAPMessageType.ACK);

                    return true;
                } else {
                    if (BuildConfig.DEBUG) { //For no slowing prod version, so many logs, what don't showing anymore
                        LogHelper.v("ARQ: Receive " + incomingMessage.getHexToken() + " in progress, responding with ACK continued, received: "
                                + receiveState.getDataSize());
                    }
                    ackMessage.setCode(CoAPMessageCode.CoapCodeContinue);

                    CoAPMessage originalMessage = messagePool.getSourceMessageByToken(incomingMessage.getHexToken());

                    if (originalMessage != null) {
                        if (originalMessage.hasOption(CoAPMessageOptionCode.OptionProxyURI)) {
                            ackMessage.addOption(originalMessage.getOption(CoAPMessageOptionCode.OptionProxyURI));
                        }
                        if (originalMessage.getProxy() != null) {
                            ackMessage.setProxy(originalMessage.getProxy());
                        }
                    }

                    client.send(ackMessage, null);
                }
                break;
            case NON:
                LogHelper.e("ARQ: NON message received");
                throw new RuntimeException("ARQ: NON message received");
        }
        incomingMessage.removeOption(blockOptionCode);
        return false;
    }

    private void didTransmit(int blockNumber, String token) {
        LogHelper.v("ARQ: did transmit block = " + blockNumber + " for token = " + token);
        SendState sendState = sendStates.get(token);
        if (sendState != null) {
            sendState.didTransmit(blockNumber);
            sendMoreData(token);
        }
    }

    private void sendMoreData(String token) {
        SendState state = sendStates.get(token);
        Block block;
        if (state != null) {
            while ((block = state.popBlock()) != null) {
                LogHelper.v("ARQ: did pop block number = " + block.getNumber());
                send(block, state.getOriginalMessage(), token, state);
                state.incrementNumberOfMessage();
                sendStates.put(token, state);
            }
        }
    }

    private void send(Block block, CoAPMessage originalMessage, String token, SendState state) {
        CoAPMessage blockMessage = new CoAPMessage(originalMessage.getType(), originalMessage.getCode());
        blockMessage.setOptions(originalMessage.getOptions());

        CoAPMessageOptionCode blockCode = originalMessage.isRequest() ? CoAPMessageOptionCode.OptionBlock1 : CoAPMessageOptionCode.OptionBlock2;
        CoAPMessageOption blockOption = new CoAPMessageOption(blockCode, block.toInt());
        blockMessage.addOption(blockOption);

        CoAPMessageOption srOption = new CoAPMessageOption(CoAPMessageOptionCode.OptionSelectiveRepeatWindowSize, state.getWindowSize());
        blockMessage.addOption(srOption);

        blockMessage.setToken(Hex.decodeHex(token.toCharArray()));
        blockMessage.setPayload(new CoAPMessagePayload(block.getData()));
        blockMessage.setURI(originalMessage.getURI());
        if (originalMessage.hasOption(CoAPMessageOptionCode.OptionProxyURI)) {
            blockMessage.addOption(originalMessage.getOption(CoAPMessageOptionCode.OptionProxyURI));
        }
        if (originalMessage.getProxy() != null) {
            blockMessage.setProxy(originalMessage.getProxy());
        }
        blockMessage.setResendHandler(state);
        client.send(blockMessage, new CoAPHandler() {
            @Override
            public void onMessage(CoAPMessage message, String error) {
                if (error != null) {
                    LogHelper.v("Block number = " + block.getNumber() + " failed " + error);
                    fail(token);
                    return;
                }
                LogHelper.v("Block number = " + block.getNumber() + " sent");
            }

            @Override
            public void onAckError(String error) {
                LogHelper.v("Block number = " + block.getNumber() + " failed");
                fail(token);
            }
        });

    }

    private void fail(String token) {
        LogHelper.v("ARQ: fail to transfer for token = " + token);
        SendState sendState = sendStates.get(token);
        if (sendState != null) {
            sendState.onError(client.getMessageDeliveryInfo(sendState.getOriginalMessage()));
            sendStates.remove(token);
        }
        receiveStates.remove(token);
    }

    @Override
    public boolean onSend(CoAPMessage message, Reference<InetSocketAddress> receiverAddressReference) {
        if (hasWindowSizeOption(message) && !isStartMixingModeMessage(message)) {
            return true;
        }

        if (!isBiggerThenCanBeTransferedBySingleBlock(message) || message.getToken() == null) {
            return true;
        }


        String token = Hex.encodeHexString(message.getToken());
        CoAPMessagePayload payload = message.getPayload();

        LogHelper.v("ARQ: removing original message " + message.getId() + " from pool");
        messagePool.remove(message);
        CoAPMessage originalMessage = null;
        CoAPMessage ackMessage = null;
        switch (message.getType()) {
            case ACK:
            case RST:
                originalMessage = new CoAPMessage(message);
                originalMessage.setType(CoAPMessageType.CON);
                originalMessage.removeOption(CoAPMessageOptionCode.OptionBlock1);
                CoAPMessage.convertToEmptyAck(message, receiverAddressReference.get());
                ackMessage = message;
                ackMessage.addOption(new CoAPMessageOption(CoAPMessageOptionCode.OptionSelectiveRepeatWindowSize, WINDOW_SIZE));
                LogHelper.d(
                        "ARQ: Send empty ack, id " + ackMessage.getId() + " " +
                                "payload: '" + ackMessage + " " +
                                "destination host: " + ackMessage.getURI() + " " +
                                "type: " + ackMessage.getType() + " " +
                                "code: " + ackMessage.getCode().name() + " " +
                                "token: " + Hex.encodeHexString(ackMessage.getToken()) + " " +
                                "options: " + MessageHelper.getMessageOptionsString(ackMessage));
                break;
            case NON:
                originalMessage = new CoAPMessage(message);
                originalMessage.setType(CoAPMessageType.CON);
                originalMessage.removeOption(CoAPMessageOptionCode.OptionBlock1);
                break;
            case CON:
                originalMessage = new CoAPMessage(message);
                break;
        }
        SendState sendState = new SendState(DataFactory.create(payload.content), WINDOW_SIZE, MAX_PAYLOAD_SIZE, originalMessage);
        sendStates.put(token, sendState);
        sendMoreData(token);

        LogHelper.v("ARQ: split message " + message.getId() + " to values. Sending payload = " + payload.content.length);

        return ackMessage != null;
    }

    private boolean isStartMixingModeMessage(CoAPMessage message) {
        return !message.getCode().isRequest()
                && message.hasOption(CoAPMessageOptionCode.OptionBlock1)
                && !message.hasOption(CoAPMessageOptionCode.OptionBlock2)
                && isBiggerThenCanBeTransferedBySingleBlock(message);
    }

    private boolean hasWindowSizeOption(CoAPMessage message) {
        return message.hasOption(CoAPMessageOptionCode.OptionSelectiveRepeatWindowSize);
    }

    private boolean isBiggerThenCanBeTransferedBySingleBlock(CoAPMessage message) {
        return message.getPayload() != null &&
                message.getPayload().content != null &&
                message.getPayload().content.length > MAX_PAYLOAD_SIZE;
    }

    public LoggableState getArqReceivingStateForToken(final byte[] token) {
        return receiveStates.get(Hex.encodeHexString(token));
    }

}

File: app/src/main/java/com/ndmsystems/coala/layers/arq/ArqLayer.kt
package com.ndmsystems.coala.layers.arq

import com.ndmsystems.coala.helpers.logging.LogHelper
import com.ndmsystems.coala.helpers.logging.LogKeys
import com.ndmsystems.coala.BuildConfig
import com.ndmsystems.coala.CoAPClient
import com.ndmsystems.coala.CoAPHandler
import com.ndmsystems.coala.CoAPMessagePool
import com.ndmsystems.coala.LayersStack
import com.ndmsystems.coala.helpers.Hex.decodeHex
import com.ndmsystems.coala.helpers.Hex.encodeHexString
import com.ndmsystems.coala.helpers.MessageHelper.getMessageOptionsString
import com.ndmsystems.coala.layers.ReceiveLayer
import com.ndmsystems.coala.layers.SendLayer
import com.ndmsystems.coala.layers.arq.data.DataFactory
import com.ndmsystems.coala.layers.arq.states.LoggableState
import com.ndmsystems.coala.layers.arq.states.ReceiveState
import com.ndmsystems.coala.layers.arq.states.SendState
import com.ndmsystems.coala.message.CoAPMessage
import com.ndmsystems.coala.message.CoAPMessageCode
import com.ndmsystems.coala.message.CoAPMessageOption
import com.ndmsystems.coala.message.CoAPMessageOptionCode
import com.ndmsystems.coala.message.CoAPMessagePayload
import com.ndmsystems.coala.message.CoAPMessageType
import com.ndmsystems.coala.utils.Reference
import net.jodah.expiringmap.ExpirationPolicy
import net.jodah.expiringmap.ExpiringMap
import java.net.InetSocketAddress
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.TimeUnit

class ArqLayer(
    private val client: CoAPClient,
    private val messagePool: CoAPMessagePool,
    /** How long a half-received transfer is kept; a parameter so tests can reach expiry. */
    receiveStateTtlMillis: Long = 10L * 60 * 1000
) : ReceiveLayer, SendLayer {
    private val receiveStates: MutableMap<String, ReceiveState> = ExpiringMap.builder()
        .expirationPolicy(ExpirationPolicy.CREATED)
        .expiration(receiveStateTtlMillis, TimeUnit.MILLISECONDS)
        .build()
    private val sendStates: MutableMap<String, SendState> = ConcurrentHashMap()
    private fun isBlockedMessage(message: CoAPMessage): Boolean {
        return message.hasOption(CoAPMessageOptionCode.OptionBlock1) ||
                message.hasOption(CoAPMessageOptionCode.OptionBlock2)
    }

    private fun isAboutArq(message: CoAPMessage): Boolean {
        return message.hasOption(CoAPMessageOptionCode.OptionSelectiveRepeatWindowSize)
    }

    private fun getBlock1(message: CoAPMessage): Block {
        val data = message.payload?.content
        return Block(
            message.getOption(CoAPMessageOptionCode.OptionBlock1)?.value as Int,
            data
        )
    }

    private fun getBlock2(message: CoAPMessage): Block {
        val data = message.payload?.content
        return Block(
            message.getOption(CoAPMessageOptionCode.OptionBlock2)?.value as Int,
            data
        )
    }

    override fun onReceive(message: CoAPMessage, senderAddressReference: Reference<InetSocketAddress>): LayersStack.LayerResult {
        if (!isAboutArq(message)) {
            return LayersStack.LayerResult(true)
        }
        if (message.code == CoAPMessageCode.CoapCodeEmpty && message.type == CoAPMessageType.ACK && !message.hasOption(CoAPMessageOptionCode.OptionBlock1)) { //For block1 block2 mixed mode, message ending block1 start block2, like Received data from Peer, id 51486, payload:'', address: 138.197.191.160:5683 type: ACK code: CoapCodeEmpty path: null schema: coap token 951e67cdc72af162
            //Options: OptionBlock1 : 'com.ndmsystems.coala.layers.arq.Block@a046fab'(8174) OptionSelectiveRepeatWindowSize : '70', address: 138.197.191.160:5683 type: CON code: CoapCodeContent path: null schema: coap token 951e67cdc72af162
            //Options: OptionBlock2 : 'com.ndmsystems.coala.layers.arq.Block@72bd608'(30) OptionSelectiveRepeatWindowSize : '300'
            messagePool.setNoNeededSending(message)
            return LayersStack.LayerResult(false)
        }
        if (!isBlockedMessage(message)) {
            return LayersStack.LayerResult(true)
        }
        if (message.token == null) {
            sendResetMessage(message, senderAddressReference.get())
        }
        val windowSize = message.getOption(CoAPMessageOptionCode.OptionSelectiveRepeatWindowSize)!!.value as Int
        val ackMessage = CoAPMessage.ackTo(message, senderAddressReference.get(), CoAPMessageCode.CoapCodeEmpty)
        if (message.hasOption(CoAPMessageOptionCode.OptionBlock1)) {
            val block = getBlock1(message)
            return process(message, CoAPMessageOptionCode.OptionBlock1, block, windowSize, ackMessage)
        }
        if (message.hasOption(CoAPMessageOptionCode.OptionBlock2)) {
            val block = getBlock2(message)
            return process(message, CoAPMessageOptionCode.OptionBlock2, block, windowSize, ackMessage)
        }
        return LayersStack.LayerResult(true)
    }

    private fun sendResetMessage(incomingMessage: CoAPMessage, fromAddress: InetSocketAddress) {
        val resetMessage = CoAPMessage.resetTo(incomingMessage, fromAddress)
        client.send(resetMessage, null)
    }

    private fun process(
        incomingMessage: CoAPMessage,
        blockOptionCode: CoAPMessageOptionCode,
        block: Block,
        windowSize: Int,
        ackMessage: CoAPMessage
    ): LayersStack.LayerResult {
        val mutableIncomingMessage: CoAPMessage = incomingMessage
        val token = encodeHexString(mutableIncomingMessage.token)
        when (mutableIncomingMessage.type) {
            CoAPMessageType.ACK, CoAPMessageType.RST -> {
                // Transmit ACK
                didTransmit(block.number, token)
                messagePool.remove(mutableIncomingMessage)

                val sendState = sendStates[token]
                if (sendState != null && sendState.isCompleted) {
                    LogHelper.v(
                        "ARQ: sending completed, pushing the original message to the pool",
                        mapOf(LogKeys.COAP_MESSAGE_ID to sendState.originalMessage.id)
                    )

                    // Completion should depend on response code, not only CoapCodeEmpty
                    if (mutableIncomingMessage.code == CoAPMessageCode.CoapCodeContinue) {
                        // Continuation — keep state until the final response
                        return LayersStack.LayerResult(false)
                    } else {
                        // Final response — state can be removed
                        val originalMessage = sendState.originalMessage
                        messagePool.add(originalMessage)
                        messagePool.setNoNeededSending(originalMessage)
                        sendStates.remove(token)
                        return LayersStack.LayerResult(true)
                    }
                }
            }

            CoAPMessageType.CON -> {
                // Receive CON
                val payload = mutableIncomingMessage.payload
                if (payload == null) {
                    LogHelper.e("ARQ: payload expected", mapOf(LogKeys.COAP_TOKEN to token))
                    throw RuntimeException("ARQ: payload expected for token = $token")
                }
                var receiveState = receiveStates[token]
                if (receiveState == null) {
                    LogHelper.v("ARQ: creating ReceiveState", mapOf(LogKeys.COAP_TOKEN to token))
                    receiveState = ReceiveState(mutableIncomingMessage)
                    receiveStates[token] = receiveState
                }
                receiveState.didReceiveBlock(block, mutableIncomingMessage.code)
                ackMessage.addOption(CoAPMessageOption(blockOptionCode, block.toInt()))
                ackMessage.addOption(CoAPMessageOption(CoAPMessageOptionCode.OptionSelectiveRepeatWindowSize, windowSize))
                if (receiveState.isTransferCompleted) {
                    //receiveStates.remove(token);
                    val originalMessage = messagePool.getSourceMessageByToken(mutableIncomingMessage.hexToken)
                    if (originalMessage != null) {
                        LogHelper.v(
                            "ARQ: receive completed, passing the original message along",
                            mapOf(
                                LogKeys.COAP_TOKEN to mutableIncomingMessage.hexToken,
                                LogKeys.BYTES to receiveState.dataSize,
                                LogKeys.COAP_MESSAGE_ID to originalMessage.id,
                                "original_token" to encodeHexString(originalMessage.token)
                            )
                        )
                        if (originalMessage.hasOption(CoAPMessageOptionCode.OptionProxyURI)) {
                            ackMessage.addOption(originalMessage.getOption(CoAPMessageOptionCode.OptionProxyURI)!!)
                        }
                        originalMessage.proxy?.let {
                            ackMessage.setProxy(it)
                            originalMessage.getProxySecurityId()?.let {
                                ackMessage.setProxySecurityId(it)
                            }
                        }
                        mutableIncomingMessage.id = originalMessage.id
                    } else {
                        LogHelper.v(
                            "ARQ: receive completed with originalMessage = null",
                            mapOf(
                                LogKeys.COAP_TOKEN to mutableIncomingMessage.hexToken,
                                LogKeys.BYTES to receiveState.dataSize
                            )
                        )
                    }
                    ackMessage.code = CoAPMessageCode.CoapCodeEmpty
                    client.send(ackMessage, null)

                    //mutableIncomingMessage = messagePool.getSourceMessageByToken(mutableIncomingMessage.getHexToken());
                    mutableIncomingMessage.payload = CoAPMessagePayload(receiveState.data ?: ByteArray(0))
                    if (blockOptionCode == CoAPMessageOptionCode.OptionBlock2) {
                        mutableIncomingMessage.code = CoAPMessageCode.CoapCodeContent
                        mutableIncomingMessage.type = CoAPMessageType.ACK
                    }
                    return LayersStack.LayerResult(true, mutableIncomingMessage)
                } else {
                    if (BuildConfig.DEBUG) { //For no slowing prod version, so many logs, what don't showing anymore
                        LogHelper.v(
                            "ARQ: receive in progress, responding with ACK continued",
                            mapOf(
                                LogKeys.COAP_TOKEN to mutableIncomingMessage.hexToken,
                                LogKeys.BYTES to receiveState.dataSize
                            )
                        )
                    }
                    ackMessage.code = CoAPMessageCode.CoapCodeContinue
                    val originalMessage = messagePool.getSourceMessageByToken(mutableIncomingMessage.hexToken)
                    if (originalMessage != null) {
                        if (originalMessage.hasOption(CoAPMessageOptionCode.OptionProxyURI)) {
                            ackMessage.addOption(originalMessage.getOption(CoAPMessageOptionCode.OptionProxyURI)!!)
                        }
                        originalMessage.proxy?.let { ackMessage.setProxy(it) }
                    }
                    client.send(ackMessage, null)
                }
            }

            CoAPMessageType.NON -> {
                LogHelper.e("ARQ: NON message received")
                throw RuntimeException("ARQ: NON message received")
            }
        }
        mutableIncomingMessage.removeOption(blockOptionCode)
        return LayersStack.LayerResult(false, mutableIncomingMessage)
    }

    private fun didTransmit(blockNumber: Int, token: String) {
        LogHelper.v(
            "ARQ: did transmit block",
            mapOf("block_number" to blockNumber, LogKeys.COAP_TOKEN to token)
        )
        val sendState = sendStates[token]
        if (sendState != null) {
            sendState.didTransmit(blockNumber)
            sendMoreData(token)
        }
    }

    private fun sendMoreData(token: String) {
        val state = sendStates[token]
        var block: Block?
        if (state != null) {
            while (state.popBlock().also { block = it } != null) {
                LogHelper.v("ARQ: did pop block", mapOf("block_number" to block!!.number))
                send(block!!, state.originalMessage, token, state)
                state.incrementNumberOfMessage()
                sendStates[token] = state
            }
        }
    }

    private fun send(block: Block, originalMessage: CoAPMessage?, token: String, state: SendState) {
        val blockMessage = CoAPMessage(originalMessage!!.type, originalMessage.code)
        blockMessage.setOptions(originalMessage.getOptions())
        val blockCode = if (originalMessage.isRequest) CoAPMessageOptionCode.OptionBlock1 else CoAPMessageOptionCode.OptionBlock2
        val blockOption = CoAPMessageOption(blockCode, block.toInt())
        blockMessage.addOption(blockOption)
        val srOption = CoAPMessageOption(CoAPMessageOptionCode.OptionSelectiveRepeatWindowSize, state.windowSize)
        blockMessage.addOption(srOption)
        blockMessage.token = decodeHex(token.toCharArray())
        blockMessage.payload = CoAPMessagePayload(block.data ?: ByteArray(0))
        blockMessage.setURI(originalMessage.getURI())
        blockMessage.addChecksumOnSend = originalMessage.addChecksumOnSend
        if (originalMessage.hasOption(CoAPMessageOptionCode.OptionProxyURI)) {
            blockMessage.addOption(originalMessage.getOption(CoAPMessageOptionCode.OptionProxyURI)!!)
        }
        originalMessage.proxy?.let { blockMessage.setProxy(it) }
        blockMessage.resendHandler = state
        client.send(blockMessage, object : CoAPHandler {
            override fun onMessage(message: CoAPMessage, error: String?) {
                if (error != null) {
                    LogHelper.v(
                        "ARQ: block failed",
                        mapOf("block_number" to block.number, LogKeys.ERROR to error)
                    )
                    fail(token)
                    return
                }
                LogHelper.v("ARQ: block sent", mapOf("block_number" to block.number))
            }

            override fun onAckError(error: String) {
                LogHelper.v("ARQ: block failed on ack", mapOf("block_number" to block.number))
                fail(token)
            }
        })
    }

    private fun fail(token: String) {
        LogHelper.v("ARQ: fail to transfer", mapOf(LogKeys.COAP_TOKEN to token))
        val sendState = sendStates[token]
        if (sendState != null) {
            sendState.onError(client.getMessageDeliveryInfo(sendState.originalMessage))
            sendStates.remove(token)
        }
        receiveStates.remove(token)
    }

    override fun onSend(message: CoAPMessage, receiverAddressReference: Reference<InetSocketAddress>): LayersStack.LayerResult {
        if (hasWindowSizeOption(message) && !isStartMixingModeMessage(message)) {
            return LayersStack.LayerResult(true)
        }
        if (!isBiggerThenCanBeTransferedBySingleBlock(message) || message.token == null) {
            return LayersStack.LayerResult(true)
        }
        val token = encodeHexString(message.token)
        val payload = message.payload
        LogHelper.v("ARQ: removing the original message from the pool", mapOf(LogKeys.COAP_MESSAGE_ID to message.id))
        messagePool.remove(message)
        var originalMessage: CoAPMessage? = null
        var ackMessage: CoAPMessage? = null
        when (message.type) {
            CoAPMessageType.ACK, CoAPMessageType.RST -> {
                originalMessage = CoAPMessage(message)
                originalMessage.type = CoAPMessageType.CON
                originalMessage.removeOption(CoAPMessageOptionCode.OptionBlock1)
                CoAPMessage.convertToEmptyAck(message, receiverAddressReference.get())
                ackMessage = message
                ackMessage.addOption(CoAPMessageOption(CoAPMessageOptionCode.OptionSelectiveRepeatWindowSize, WINDOW_SIZE))
                LogHelper.d(
                    "ARQ: send empty ack",
                    mapOf(
                        LogKeys.COAP_MESSAGE_ID to ackMessage.id,
                        LogKeys.PAYLOAD to ackMessage.toString(),
                        LogKeys.URL to ackMessage.getURI(),
                        "coap_type" to ackMessage.type.toString(),
                        "coap_code" to ackMessage.code.name,
                        LogKeys.COAP_TOKEN to encodeHexString(ackMessage.token),
                        "options" to getMessageOptionsString(ackMessage)
                    )
                )
            }

            CoAPMessageType.NON -> {
                originalMessage = CoAPMessage(message)
                originalMessage.type = CoAPMessageType.CON
                originalMessage.removeOption(CoAPMessageOptionCode.OptionBlock1)
            }

            CoAPMessageType.CON -> originalMessage = CoAPMessage(message)
        }
        val sendState = SendState(DataFactory.create(payload?.content ?: ByteArray(0)), WINDOW_SIZE, MAX_PAYLOAD_SIZE, originalMessage)
        sendStates[token] = sendState
        sendMoreData(token)
        LogHelper.v(
            "ARQ: split message into blocks",
            mapOf(LogKeys.COAP_MESSAGE_ID to message.id, LogKeys.BYTES to payload?.content?.size)
        )
        return LayersStack.LayerResult(ackMessage != null)
    }

    private fun isStartMixingModeMessage(message: CoAPMessage): Boolean {
        return (!message.code.isRequest
                && message.hasOption(CoAPMessageOptionCode.OptionBlock1)
                && !message.hasOption(CoAPMessageOptionCode.OptionBlock2)
                && isBiggerThenCanBeTransferedBySingleBlock(message))
    }

    private fun hasWindowSizeOption(message: CoAPMessage): Boolean {
        return message.hasOption(CoAPMessageOptionCode.OptionSelectiveRepeatWindowSize)
    }

    private fun isBiggerThenCanBeTransferedBySingleBlock(message: CoAPMessage): Boolean {
        return message.payload != null && message.payload!!.content.size > MAX_PAYLOAD_SIZE
    }

    fun getArqReceivingStateForToken(token: ByteArray?): LoggableState? {
        return receiveStates[encodeHexString(token)]
    }

    companion object {
        private const val WINDOW_SIZE = 70
        private const val MAX_PAYLOAD_SIZE = 1024
    }

    override fun onStop() {
        receiveStates.clear()
        sendStates.clear()
    }
}
